### PR TITLE
 Release a pip installable package

### PR DESCRIPTION
**Pull request by LukeLIN-web**
See  https://packaging.python.org/en/latest/tutorials/packaging-projects/ 
The following command is enough:
```
python3 -m pip install --upgrade build
python3 -m build
python3 -m pip install --upgrade twine
python3 -m twine upload dist/*
```

You can add more information, such as the author's names and emails.
I installed it from my pypi  [link](https://pypi.org/project/flexgen/ )    It works great for me.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -5,9 +5,14 @@
...
+    name="flexgen",
+    version='1.0',
+    description='Running large language models like OPT-175B/GPT-3 on a single GPU. Focusing on high-throughput large-batch generation.',
+    python_requires='>=3.7',
+    install_requires=install_requires,
+    packages=find_packages(exclude=[
+        "benchmark", "experiments", "playground", "scripts",
+        "third_party",
+    ]),
+)
```
Comment (Borda) [inline setup.py:18]:
```suggestion
setup()
```

File: pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+readme = "README.md"
+requires-python = ">=3.7"
+classifiers = [
+    "Programming Language :: Python :: 3",
+    "License :: OSI Approved :: MIT License",
+]
+
+[project.urls]
+"Homepage" = "https://github.com/FMInference/FlexGen"
+"Bug Tracker" = "https://github.com/FMInference/FlexGen/issues"
```
Comment (Borda) [inline pyproject.toml:19]:
```suggestion
"Bug Tracker" = "https://github.com/FMInference/FlexGen/issues"

[tool.setuptools.packages.find]
exclude = ["benchmark", "experiments", "playground", "scripts", "third_party"]
```

File: pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+build-backend = "setuptools.build_meta"
+
+
+[project]
+name = "flexgen"
+version = "1.0"
+description = "Running large language models like OPT-175B/GPT-3 on a single GPU. Focusing on high-throughput large-batch generation."
+readme = "README.md"
+requires-python = ">=3.7"
+classifiers = [
```
Comment (Borda) [inline pyproject.toml:12]:
```suggestion
install_requires = ["numpy", "tqdm", "pulp", "attrs", "transformers>=4.24", "torch>=1.12"]
classifiers = [
```

Comment (LukeLIN-web) [reply pyproject.toml:13]: It leads to 
```
 ValueError: invalid pyproject.toml config: `project`.
      configuration error: `project` must not contain {'install_requires'} properties
      [end of output]
  ```

File: setup.py (hunk 1 above)
Comment (LukeLIN-web) [reply setup.py]: Why not contain in setup?

File: pyproject.toml (hunk 3 above)
Comment (Borda) [reply pyproject.toml:12]: the proper name is `dependencies`, updated

**Comment by maxsu** [comment]
Note, we can:

1. Migrate `setup_depends` from `setup.py` into the `buildsystem.requires` key in `pyproject.toml`. 
2. Delete `setup.py`


#### Reasoning

The rationale for dropping `setup_depends` is written in [pep 518, section 2](https://peps.python.org/pep-0518/#rationale). This is now endorsed by the [python packaging tutorial](https://packaging.python.org/en/latest/tutorials/packaging-projects/#creating-pyproject-toml) and [pip v23.0.1 manual](https://pip.pypa.io/en/stable/reference/build-system/pyproject-toml/#build-time-dependencies). After this, `setup.py` is completely empty, so it's safe to sweep it🧹

**Comment by Ying1123** [comment]
This looks good. I will review it in detail later. Could you transfer the ownership of this package in pypi to my account  sqy1415@gmail.com ?

**Comment by LukeLIN-web** [comment]
> This looks good. I will review it in detail later. Could you transfer the ownership of this package in pypi to my account [sqy1415@gmail.com](mailto:sqy1415@gmail.com) ? My username is ying1123.

Invitation sent to 'ying1123'

**Comment by Ying1123** [comment]
> > This looks good. I will review it in detail later. Could you transfer the ownership of this package in pypi to my account [sqy1415@gmail.com](mailto:sqy1415@gmail.com) ? My username is ying1123.
> 
> Invitation sent to 'ying1123'

Thanks a lot, @LukeLIN-web! I have accepted the invitation and will return with the pip release tomorrow.

**Comment by LukeLIN-web** [comment]
> Note, we can:
> 
> 1. Migrate `setup_depends` from `setup.py` into the `buildsystem.requires` key in `pyproject.toml`.
> 2. Delete `setup.py`
> 
> #### Reasoning
> The rationale for dropping `setup_depends` is written in [pep 518](https://peps.python.org/pep-0518/#rationale). This is now endorsed by the [python packaging tutorial](https://packaging.python.org/en/latest/tutorials/packaging-projects/#creating-pyproject-toml) and [pip v23.0.1 manual](https://pip.pypa.io/en/stable/reference/build-system/pyproject-toml/#build-time-dependencies). After this, `setup.py` is an empty `setup()` invocation, so it's safe to sweep🧹

Done.

**Comment by anirrudh** [comment]
I'd be happy to package this for conda in addition to pip if that is of any interest!

**Comment by maxsu** [comment]
@LukeLIN-web we use setup.py due to pip's dependency on setuptools. But TIL setuptools fully embraced pyproject.toml early last year: 

https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html
https://github.com/pypa/setuptools/commits/main/docs/userguide/pyproject_config.rst

I've noticed project config is simpler when it lives in a toml. Many in the PY community have worked towards this over the past few years, and I'm happy to finally get rid of setup.py in my projects.

I now think of `setup.py` as a home for very specialized setup logic - testing GPU availability for instance. Most projects won't need it at all. When they do need it, the file can be pure logic, since the metadata lives elsewhere.

</off-topic>

**Comment by Ying1123** [comment]
Thanks. It is merged!

**Pull request merged**